arbor/include/arbor/network.hpp

```diff
@@ -0,0 +1,303 @@
+
+private:
+    network_value(std::shared_ptr<network_value_impl> impl);
+
+    friend std::shared_ptr<network_value_impl> thingify(network_value v,
+        const network_label_dict& dict);
+
+    std::shared_ptr<network_value_impl> impl_;
+};
+
+ARB_ARBOR_API inline network_value operator+(network_value a, network_value b) {
+    return network_value::add(std::move(a), std::move(b));
+}
+
+ARB_ARBOR_API inline network_value operator-(network_value a, network_value b) {
+    return network_value::sub(std::move(a), std::move(b));
+}
+
+ARB_ARBOR_API inline network_value operator*(network_value a, network_value b) {
+    return network_value::mul(std::move(a), std::move(b));
+}
+
+ARB_ARBOR_API inline network_value operator/(network_value a, network_value b) {
+    return network_value::div(std::move(a), std::move(b));
+}
+
+ARB_ARBOR_API inline network_value operator+(network_value a) { return a; }
+
+ARB_ARBOR_API inline network_value operator-(network_value a) {
+    return network_value::mul(-1.0, std::move(a));
+}
+
+class ARB_SYMBOL_VISIBLE network_selection {
+public:
+    using custom_func_type = std::function<bool(const network_connection_info&)>;
+
+    network_selection() { *this = network_selection::none(); }
+
+    // Select all
+    static network_selection all();
+
+    // Select none
+    static network_selection none();
+
+    // Named selection in the network label dictionary
+    static network_selection named(std::string name);
+
+    // Only select connections between different cells
+    static network_selection inter_cell();
+
+    // Select connections with the given source cell kind
+    static network_selection source_cell_kind(cell_kind kind);
+
+    // Select connections with the given destination cell kind
+    static network_selection destination_cell_kind(cell_kind kind);
+
+    // Select connections with the given source label
+    static network_selection source_label(std::vector<cell_tag_type> labels);
+
+    // Select connections with the given destination label
+    static network_selection destination_label(std::vector<cell_tag_type> labels);
+
+    // Select connections with source cells matching the indices in the list
+    static network_selection source_cell(std::vector<cell_gid_type> gids);
+
+    // Select connections with source cells matching the indices in the range
+    static network_selection source_cell(gid_range range);
+
+    // Select connections with destination cells matching the indices in the list
+    static network_selection destination_cell(std::vector<cell_gid_type> gids);
+
+    // Select connections with destination cells matching the indices in the range
+    static network_selection destination_cell(gid_range range);
+
+    // Select connections that form a chain, such that source cell "i" is connected to the destination cell "i+1"
+    static network_selection chain(std::vector<cell_gid_type> gids);
```

Comment:
I think this might become quite complicated to understand and it only solved the case, where the chain crossed from the maximum index to 0, which is not difficult to express in different ways.